bigtable/client/row_range.cc

```diff
@@ -0,0 +1,446 @@
+ *   of type CLOSED. Both should be treated equal.
+ */
+bool RowRange::operator==(RowRange const& rhs) const {
+  std::string start_key{GetStartKey()}, end_key{GetEndKey()},
+      rhs_start_key{rhs.GetStartKey()}, rhs_end_key{rhs.GetEndKey()};
+  KeyType start_type{GetStartKeyType()}, end_type{GetEndKeyType()},
+      rhs_start_type{rhs.GetStartKeyType()}, rhs_end_type{rhs.GetEndKeyType()};
+
+  return ((this->IsEmpty() && rhs.IsEmpty()) ||
+          ((start_key == rhs_start_key &&
+            /* If key is empty, type is 'don't care'*/
+            (start_key.empty() || start_type == rhs_start_type)) &&
+           (end_key == rhs_end_key && end_type == rhs_end_type)));
+}
+
+// The intersection of two ranges (useful for retries).
+RowRange RowRange::intersect(RowRange const& lhs, RowRange const& rhs) {
+  std::string lhs_start_key{lhs.GetStartKey()},
+              lhs_end_key{lhs.GetEndKey()},
+              rhs_start_key{rhs.GetStartKey()},
+              rhs_end_key{rhs.GetEndKey()};
+  KeyType lhs_start_type{lhs.GetStartKeyType()},
+          lhs_end_type{lhs.GetEndKeyType()},
+          rhs_start_type{rhs.GetStartKeyType()},
+          rhs_end_type{rhs.GetEndKeyType()};
+  std::string result_start_key{lhs_start_key}, result_end_key{lhs_end_key};
+  KeyType result_start_type{lhs_start_type}, result_end_type{lhs_end_type};
+
+  // Fix the start key and type: pick the max value.
+  FixKeyPointAndTypeIntersect(PositionType::START, result_start_key,
+                              result_start_type, rhs_start_key, rhs_start_type);
+  // Fix end key and type: Pick the min value.
+  FixKeyPointAndTypeIntersect(PositionType::END, result_end_key,
+                              result_end_type, rhs_end_key, rhs_end_type);
+
+  // Avoid making  result_end_key  smaller than result_start_key.
+  if (result_end_key < result_start_key) {
+    result_end_key = result_start_key;
+  }
+
+  RowRange tmp;
+  SetProtoBufRangePoint(PositionType::START, tmp, result_start_key,
+                        result_start_type);
+  SetProtoBufRangePoint(PositionType::END, tmp, result_end_key,
+                        result_end_type);
+  return tmp;
+}
+
+/**
+ * Two scenarios for empty range:
+ * 1. For normal input values (of start and end), an empty range satisfies all
+ * the properties below:
+ *   - either start or end key point is open,
+ *   - start key == end key,
+ *   - end key is 'set' (i.e., a finite range).
+ * 2. For anomalous inputs ('end' gets a value lower than that of 'start'),
+ * a range is empty if:
+ *    - the end is set and end < start.
+ *
+ *  We exclude open ranges like, for example, (x, Successor(x)) from
+ *  consideration, as they could be prefix ranges with x being the prefix
+ *  of a row key.
+ */
+bool RowRange::IsEmpty() const {
+  KeyType start_type{GetStartKeyType()}, end_type{GetEndKeyType()};
+  std::string start_key{GetStartKey()}, end_key{GetEndKey()};
+
+  return (/* Case 1. */
```

Comment:
Please break that long complex boolean expression into a couple of `if()` statements.